gptel-bedrock.el

```diff
@@ -0,0 +1,582 @@
+    (mistral-small-2402          . "mistral.mistral-small-2402-v1:0")
+    (llama-3-8b                  . "meta.llama3-8b-instruct-v1:0")
+    (llama-3-70b                 . "meta.llama3-70b-instruct-v1:0")
+    (llama-3-1-8b                . "meta.llama3-1-8b-instruct-v1:0")
+    (llama-3-1-70b               . "meta.llama3-1-70b-instruct-v1:0")
+    (llama-3-1-405b              . "meta.llama3-1-405b-instruct-v1:0")
+    (llama-3-2-1b                . "meta.llama3-2-1b-instruct-v1:0")
+    (llama-3-2-3b                . "meta.llama3-2-3b-instruct-v1:0")
+    (llama-3-2-11b               . "meta.llama3-2-11b-instruct-v1:0")
+    (llama-3-2-90b               . "meta.llama3-2-90b-instruct-v1:0")
+    (llama-3-3-70b               . "meta.llama3-3-70b-instruct-v1:0"))
+  "Map of model name to bedrock id.
+
+IDs can be added or replaced by calling
+\(push (model-name . \"model-id\") gptel-bedrock-model-ids).")
+
+(defvar gptel--bedrock-models
+  (let ((known-ids (mapcar #'car gptel-bedrock-model-ids)))
+    (cl-remove-if-not (lambda (model) (memq (car model) known-ids)) gptel--anthropic-models))
+  "List of available AWS Bedrock models and associated properties.")
+
+(defun gptel-bedrock--get-model-id (model)
+  "Return the Bedrock model ID for MODEL."
+  (or (alist-get model gptel-bedrock-model-ids nil nil #'eq)
+      (error "Unknown Bedrock model: %s" model)))
+
+(defun gptel-bedrock--curl-args (region)
+  "Generate the curl arguments to get a bedrock request signed for use in REGION."
+  ;; https://curl.se/docs/manpage.html#--aws-sigv4
+  (cl-multiple-value-bind (key-id secret token) (gptel-bedrock--get-credentials)
+    (nconc
+     (list
+      "--user" (format "%s:%s" key-id secret)
+      "--aws-sigv4" (format "aws:amz:%s:bedrock" region)
+      "--output" "/dev/stdout") ;; Without this curl swallows the output
+     (when token
+       (list (format "-Hx-amz-security-token: %s" token))))))
+
+;;;###autoload
+(cl-defun gptel-make-bedrock
+    (name &key
+          region
+          (models gptel--bedrock-models)
+          (stream nil)
+          (protocol "https"))
+  "Register an AWS Bedrock backend for gptel with NAME.
+
+Keyword arguments:
+
+REGION - AWS region name (e.g. \"us-east-1\")
+MODELS - The list of models supported by this backend
+STREAM - Whether to use streaming responses or not."
+  (declare (indent 1))
+  (let ((host (format "bedrock-runtime.%s.amazonaws.com" region)))
+    (setf (alist-get name gptel--known-backends nil nil #'equal)
+          (gptel--make-bedrock
+           :name name
+           :host host
+           :header nil           ; x-amz-security-token is set in curl-args if needed
+           :models (gptel--process-models models)
+           :protocol protocol
+           :endpoint "" ; Url is dynamically constructed based on other args
+           :stream stream
+           :coding-system (and stream 'binary)
+           :curl-args (lambda () (gptel-bedrock--curl-args region))
```

Comment:
This will mean that the user can't add any more backend-specific Curl arguments.  For example, if they want to specify `--max-time` or `--keepalive-time` time for this backend.